File: algorithms/EvolutionaryProgramming/EP.py
# Evolutionary Programming algorithm in the Python Programming Language

# Adapted from:
# The Clever Algorithms Project: http://www.CleverAlgorithms.com
# (c) Copyright 2010 Jason Brownlee. Some Rights Reserved.
# This work is licensed under a Creative Commons Attribution-Noncommercial-Share Alike 2.5 Australia License.

import math
import random
from random import randint
import numpy as np
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt
import random


def fun(x, y):
    return x ** 2 + y ** 2


def objective_function(vector):                                                                                         # Calculate Z value (How 'high' from the min) This function contains what the minimums look like
    return ((vector[1]**2.0) + (vector[0]**2.0))



def random_vector(minmax):                                                                                              # Gives random numbers between the bounds set [-5, 5] for random vector and [0, 0.5] for strategy
    rand_vec = []
    for i in range(0, len(minmax)):
        rand_vec.append(minmax[i][0] + ((minmax[i][1] - minmax[i][0]) * random.random()))
    return rand_vec



def random_gaussian(mean=0.0, stdev=1.0):                                                                               # Standard normal distribution that is from [-3, 3]
  u1 = u2 = w = 0
  while True:
    u1 = 2 * random.random() - 1
    u2 = 2 * random.random() - 1
    w = u1 * u1 + u2 * u2
    if not (w >= 1):
        break
  w = math.sqrt((-2.0 * math.log(w)) / w)
  return (mean + (u2 * w) * stdev)




def mutate(candidate, search_space):
    child = {'vector':[], 'strategy':[]}
    for i in range(0, len(candidate['vector'])):
        s_old = candidate['strategy'][i]
        v = candidate['vector'][i] + s_old * random_gaussian()
        if v < search_space[i][0]: v = search_space[i][0]
        if v > search_space[i][1]: v = search_space[i][1]
        child['vector'].append(v)
        child['strategy'].append(s_old + random_gaussian() * abs(s_old) ** 0.5)
    return child



def tournament(candidate, population, bout_size):
  candidate['wins'] = 0
  for i in range(0, bout_size):
    other = population[randint(0,len(population)-1)]
    if candidate['fitness'] < other['fitness']: candidate['wins'] += 1



def init_population(minmax, pop_size):
    strategy = []
    for i in range(0, len(minmax)):
        strategy.append([0,  (minmax[i][1]-minmax[i][0]) * 0.05])
    pop = [dict() for x in range(0, pop_size)]
    for i in range(0, len(pop)):
        pop[i]['vector'] = random_vector(minmax)
        pop[i]['strategy'] = random_vector(strategy)
    for i in pop:
        i['fitness']= objective_function(i['vector'])
    return pop



def search (max_gens, search_space, pop_size, bout_size):
    population = init_population(search_space, pop_size)
    for i in population:
        i['fitness'] = objective_function(i['vector'])
    best = sorted(population, key=lambda i: i['fitness'])[0]

    for gen in range(0, max_gens):
        best = sorted(population, key=lambda i: i['fitness'])[0]

        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
        x = y = np.arange(-5, 5, 0.5)
        X, Y = np.meshgrid(x, y)
        zs = np.array([fun(x, y) for x, y in zip(np.ravel(X), np.ravel(Y))])
        Z = zs.reshape(X.shape)

        ax.plot_wireframe(X, Y, Z)

        ax.set_xlabel('X ')
        ax.set_ylabel('Y ')
        ax.set_zlabel('Z ')
        ax.view_init(70, 90)
        plt.plot([0], [0], [0], 'ro')
        for i in population:
            plt.plot([i['vector'][0]], [i['vector'][1]], [i['fitness']], 'go')
        plt.savefig(str(gen) + '.png')


        children = []
        for i in range(0,pop_size): children.append(mutate(population[i], search_space))
        for c in children: c['fitness'] = objective_function(c['vector'])
        children = sorted(children, key=lambda i: i['fitness'])
        if children[0]['fitness'] < best['fitness']: best = children[0]
        union = children + population
        for c in union: tournament(c, union, bout_size)
        union = sorted(union, key=lambda i: i['wins'])[::-1]
        population = union[0:pop_size]
        print (" > gen " + str(gen) + " fitness = " + str(best['fitness']) + " vector = " + str(best['vector'] ) + " strat = " + str(best['strategy'] ))


    return best



if __name__== "__main__":
    # problem configuration
    problem_size = 2
    search_space = []
    for i in range(0,problem_size): search_space.append([-5, +5])
    max_gens = 200
    pop_size = 100
    bout_size = 5
    # execute the algorithm
    best = search(max_gens, search_space, pop_size, bout_size)
    print(" done! Solution: f= " + str(best['fitness']) + " s = " + str(best['vector']))

File: algorithms/PastAlgorithms/EvolutionaryProgramming/EP.rb
# Evolutionary Programming algorithm in the Ruby Programming Language

# The Clever Algorithms Project: http://www.CleverAlgorithms.com
# (c) Copyright 2010 Jason Brownlee. Some Rights Reserved.
# This work is licensed under a Creative Commons Attribution-Noncommercial-Share Alike 2.5 Australia License.

def objective_function(vector)
  return vector.inject(0.0) {|sum, x| sum +  (x ** 2.0)}                                                                # Calculates fitness value. sums up vector[0]^2 + vector[1]^2 and that's the fitness
end

def random_vector(minmax)
  return Array.new(minmax.size) do |i|
    minmax[i][0] + ((minmax[i][1] - minmax[i][0]) * rand())
  end
end

def random_gaussian(mean=0.0, stdev=1.0)
  u1 = u2 = w = 0
  begin
    u1 = 2 * rand() - 1                                                                                                 # rand() generates float from 0 to 1.0
    u2 = 2 * rand() - 1
    w = u1 * u1 + u2 * u2                                                                                               # u1^2 + u2^2 = w
  end while w >= 1                                                                                                      # Make sure w is less than or equal to 1
  w = Math.sqrt((-2.0 * Math.log(w)) / w)                                                                               # Do calculations to w
  return mean + (u2 * w) * stdev                                                                                        # Return w with mean and stdev parameters
end

def mutate(candidate, search_space)
  child = {:vector=>[], :strategy=>[]}                                                                                  # Creates empty hash called child with 'vector' key and 'strategy' key that both have empty arrays
  candidate[:vector].each_with_index do |v_old, i|                                                                      # Iterate over candidate[:vector] which has 2 elements, v_old = value, i = index
    s_old = candidate[:strategy][i]                                                                                     # s_old = strategy old
    v = v_old + s_old * random_gaussian()                                                                               # v = strategy old + vector old * random_gaussian()
    v = search_space[i][0] if v < search_space[i][0]                                                                    # if v is less than minimum search space, v equals the minimum search space. (Contains v within)
    v = search_space[i][1] if v > search_space[i][1]                                                                    # if v is greater than maximum search space, v equals the maximum search space.
    child[:vector] << v                                                                                                 # Appends v to child hash under lab 'vector'
    child[:strategy] << s_old + random_gaussian() * s_old.abs**0.5                                                      # Appends a calculation including random_gaussian() to the strategy
  end
  return child                                                                                                          # Returns child hash
end

def tournament(candidate, population, bout_size)
  candidate[:wins] = 0                                                                                                  # Candidate is that single hash, put a new key value pair called 'wins' and initialize at 0
  bout_size.times do |i|                                                                                                # Do this loop for bout_size, in this case 5
    other = population[rand(population.size)]                                                                           # From the whole population, now with both children and parents so 200 elements since the original size was 100, pick a random index that points to a single 3 element hash and name that hash as other
    candidate[:wins] += 1 if candidate[:fitness] < other[:fitness]                                                      # if the current single candidate hash being tested has a lower fitness than the random fitness, mark it as a win in the candidate 'wins' key.
  end                                                                                                                   # Thus, a candidate can have a max of 5 wins and min of 0
end

#Why are all pop the same?
def init_population(minmax, pop_size)
  strategy = Array.new(minmax.size) do |i|                                                                              # Make a new array called strategy that holds [0, (5 - (-5) * 0.5)] problem_size times. [[], []] Nested array
    [0,  (minmax[i][1]-minmax[i][0]) * 0.05]
  end
  pop = Array.new(pop_size, {})                                                                                         # Array of size pop, 100 in this case of empty hashes
  pop.each_index do |i|                                                                                                 # For each element in pop which is a [{}, {}]
    pop[i][:vector] = random_vector(minmax)                                                                             # Run random_vector on the original search space and put it in the pop[index][:vector]
    pop[i][:strategy] = random_vector(strategy)                                                                         # Run random_vector on the strategy created earlier in this method and append it to pop[index][:strategy]
  end
  pop.each{|c| c[:fitness] = objective_function(c[:vector])}                                                            # Calculates a fitness for each vector using objective_function
  return pop
end

def search(max_gens, search_space, pop_size, bout_size)
  population = init_population(search_space, pop_size)
  population.each{|c| c[:fitness] = objective_function(c[:vector])}                                                     # Calculates fitness of each vector....again?
  best = population.sort{|x,y| x[:fitness] <=> y[:fitness]}.first                                                       # Sorts by fitness and the best is the first element: It is a hash with 'vector' 'strategy' 'fitness'
  max_gens.times do |gen|
    children = Array.new(pop_size) {|i| mutate(population[i], search_space)}                                            # Creates mutations of current pop that fall within search space and witch each hash in the population
    children.each{|c| c[:fitness] = objective_function(c[:vector])}                                                     # Calculates fitness for all child nodes
    children.sort!{|x,y| x[:fitness] <=> y[:fitness]}                                                                   # Sorts all the hashes by fitness number in ascending order from 0 = lowest, pop_size = highest
    best = children.first if children.first[:fitness] < best[:fitness]                                                  # best is now equal to the first (lowest fitness) child node if it is lower than the current best fitness. Keep in mind that best is a single 3 element hash with 'vector' = [], 'strategy' = [] and 'fitness' = float
    union = children+population                                                                                         # Combines both population and current children
    union.each{|c| tournament(c, union, bout_size)}                                                                     # Passes single hash, whole array of hashes, and bout_size which is 5 here, this will happen 200 times here and returns that hash set with 'wins' key value pair which tells how many times a random fitness in union is less than the one in that set
    union.sort!{|x,y| y[:wins] <=> x[:wins]}                                                                            # Sort by number of wins from 0 = most, union.size = least
    population = union.first(pop_size)                                                                                  # The first pop_size of the union become the new population
    puts " > gen #{gen}, fitness=#{best[:fitness]}"                                                                     # Prints out once each gen , fitness should get less and less
  end
  return best
end

if __FILE__ == $0
  # problem configuration
  problem_size = 2                                                                                                      #
  search_space = Array.new(problem_size) {|i| [-5, +5]}                                                                 # Makes array of 2 element arrays with value [-5, 5]
  # algorithm configuration
  max_gens = 200
  pop_size = 100
  bout_size = 5
  # execute the algorithm
  best = search(max_gens, search_space, pop_size, bout_size)
  puts "done! Solution: f=#{best[:fitness]}, s=#{best[:vector].inspect}"
end
